pkg/controller/node/taint_controller.go

```diff
@@ -0,0 +1,414 @@
+					tc.handleNodeUpdate(nodeUpdate)
+				default:
+					break priority
+				}
+			}
+			// After Node queue is emptied we process podUpdate.
+			tc.handlePodUpdate(podUpdate)
+		}
+	}
+}
+
+// PodUpdated is used to notify NoExecuteTaintManager about Pod changes.
+func (tc *NoExecuteTaintManager) PodUpdated(oldPod *v1.Pod, newPod *v1.Pod) {
+	var err error
+	oldTolerations := []v1.Toleration{}
+	if oldPod != nil {
+		oldTolerations, err = v1.GetPodTolerations(oldPod)
+		if err != nil {
+			glog.Errorf("Failed to get Tolerations from the old Pod: %v", err)
+			return
+		}
+	}
+	newTolerations := []v1.Toleration{}
+	if newPod != nil {
+		newTolerations, err = v1.GetPodTolerations(newPod)
+		if err != nil {
+			glog.Errorf("Failed to get Tolerations from the new Pod: %v", err)
+			return
+		}
+	}
+
+	if oldPod != nil && newPod != nil && api.Semantic.DeepEqual(oldTolerations, newTolerations) && oldPod.Spec.NodeName == newPod.Spec.NodeName {
+		return
+	}
+	select {
+	case tc.podUpdateChannel <- podUpdateItem{
+		oldPod:         oldPod,
+		newPod:         newPod,
+		oldTolerations: oldTolerations,
+		newTolerations: newTolerations,
+	}:
+	default:
+		// If our update queue is full it means that something's blocked, and we
+		// should try restarting controller.
+		panic(fmt.Errorf("NoExecuteTaintManager Pod update queue is full"))
+	}
+}
+
+// NodeUpdated is used to notify NoExecuteTaintManager about Node changes.
+func (tc *NoExecuteTaintManager) NodeUpdated(oldNode *v1.Node, newNode *v1.Node) {
+	var err error
+	oldTaints := []v1.Taint{}
+	if oldNode != nil {
+		oldTaints, err = v1.GetNodeTaints(oldNode)
+		if err != nil {
+			glog.Errorf("Failed to get Taints from the old Node: %v", err)
+			return
+		}
+	}
+	oldTaints = getNonExecuteTaints(oldTaints)
+
+	newTaints := []v1.Taint{}
+	if newNode != nil {
+		newTaints, err = v1.GetNodeTaints(newNode)
+		if err != nil {
+			glog.Errorf("Failed to get Taints from the new Node: %v", err)
+			return
+		}
+	}
+	newTaints = getNonExecuteTaints(newTaints)
+
+	if oldNode != nil && newNode != nil && api.Semantic.DeepEqual(oldTaints, newTaints) {
+		return
+	}
+	select {
+	case tc.nodeUpdateChannel <- nodeUpdateItem{
+		oldNode:   oldNode,
+		newNode:   newNode,
+		oldTaints: oldTaints,
+		newTaints: newTaints,
+	}:
+	default:
+		// If our update queue is full it means that something's blocked, and we
+		// should try restarting controller.
+		panic(fmt.Errorf("NoExecuteTaintManager Node update queue is full"))
+	}
+}
+
+func (tc *NoExecuteTaintManager) processPodOnNode(
+	podNamespacedName types.NamespacedName,
+	nodeName string,
+	tolerations []v1.Toleration,
+	taints []v1.Taint,
+	now time.Time,
+) {
+	allTolerated, usedTolerations := getMatchingTolerations(taints, tolerations)
+	if !allTolerated {
+		glog.V(2).Infof("Not all taints are tolerated after upgrade for Pod %v on %v", podNamespacedName.String(), nodeName)
+		tc.taintEvictionQueue.CancelWork(podNamespacedName.String())
```

Comment:
I don't understand it. If not all are tolerated, why we are cancelling work? Just to ensure that there is exactly one piece of work for every pod? If so, please add a comment.